server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Rights.java

```diff
@@ -0,0 +1,203 @@
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.james.mailbox.model.MailboxACL;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Multimap;
+
+public class Rights {
+    public enum Right {
+        Administer('a'),
+        Expunge('e'),
+        Insert('i'),
+        Lookup('l'),
+        Read('r'),
+        Seen('s'),
+        T_Delete('t'),
+        Write('w');
+
+        private final char imapRight;
+
+        Right(char imapRight) {
+            this.imapRight = imapRight;
+        }
+
+        @JsonValue
+        public char getImapRight() {
+            return imapRight;
+        }
+
+        public static boolean exists(char c) {
+            return Arrays.stream(values())
+                .anyMatch(right -> right.getImapRight() == c);
+        }
+
+        public static Right forChar(char c) {
+            return Arrays.stream(values())
+                .filter(right -> right.getImapRight() == c)
+                .findAny()
+                .orElseThrow(() -> new IllegalArgumentException("No matching right for '" + c + "'"));
+        }
+    }
+
+    public static class Username {
+        private final String value;
+
+        public Username(String value) {
+            this.value = value;
+        }
+
+        @JsonValue
+        public String getValue() {
+            return value;
+        }
+
+        @Override
+        public final boolean equals(Object o) {
+            if (o instanceof Username) {
+                Username username = (Username) o;
+
+                return Objects.equals(this.value, username.value);
+            }
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(value);
+        }
+    }
+
+    public static class Builder {
+        private Multimap<Username, Right> rights;
+
+        public Builder() {
+            rights = ArrayListMultimap.create();
+        }
+
+        public Builder delegateTo(Username username, Right... rights) {
+            delegateTo(username, Arrays.asList(rights));
+            return this;
+        }
+
+        public Builder delegateTo(Username username, Collection<Right> rights) {
+            this.rights.putAll(username, rights);
+            return this;
+        }
+
+        public Builder combine(Builder builder) {
+            this.rights.putAll(builder.rights);
+            return this;
+        }
+
+        public Rights build() {
+            return new Rights(rights);
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static Rights fromACL(MailboxACL acl) {
```

Comment:
this method is too complex, should be refactored

Reply:
The big "map" have been extracted